crates/sp-domains-fraud-proof/src/fraud_proof.rs

```diff
@@ -300,108 +300,89 @@ pub enum VerificationError {
-    pub bundle_index: u32,
-}
-
-impl FalseInvalidBundleEntryFraudProof {
-    pub fn new(bad_receipt_hash: ReceiptHash, domain_id: DomainId, bundle_index: u32) -> Self {
-        Self {
-            bad_receipt_hash,
-            domain_id,
-            bundle_index,
-        }
-    }
+    ReceivedInvalidInfoFromTxRangeHostFn,
 }
 
 /// Proof data specific to each *expected* invalid bundle type
 #[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
-pub enum ProofDataPerExpectedInvalidBundle {
+pub enum ProofDataPerInvalidBundleType {
     OutOfRangeTx,
 }
 
 #[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
-pub struct TrueInvalidBundleEntryFraudProof {
+pub struct InvalidBundlesFraudProof {
     pub bad_receipt_hash: ReceiptHash,
     pub domain_id: DomainId,
     pub bundle_index: u32,
     pub mismatched_extrinsic_index: u32,
     pub extrinsic_inclusion_proof: Vec<Vec<u8>>,
-    pub proof_data: ProofDataPerExpectedInvalidBundle,
+    pub is_true_invalid_fraud_proof: bool,
+    pub proof_data: ProofDataPerInvalidBundleType,
 }
 
-impl TrueInvalidBundleEntryFraudProof {
+impl InvalidBundlesFraudProof {
     pub fn new(
         bad_receipt_hash: ReceiptHash,
         domain_id: DomainId,
         bundle_index: u32,
         mismatched_extrinsic_index: u32,
         extrinsic_inclusion_proof: Vec<Vec<u8>>,
-        proof_data: ProofDataPerExpectedInvalidBundle,
+        is_true_invalid_fraud_proof: bool,
+        proof_data: ProofDataPerInvalidBundleType,
     ) -> Self {
         Self {
             bad_receipt_hash,
             domain_id,
             bundle_index,
             mismatched_extrinsic_index,
             extrinsic_inclusion_proof,
+            is_true_invalid_fraud_proof,
             proof_data,
         }
     }
 
-    pub fn matches_with_bundle_entry(&self, inboxed_bundle_entry: &InboxedBundle) -> bool {
-        if !inboxed_bundle_entry.is_invalid() {
-            return true;
-        }
-
-        let expected_bundle_validity = match self.proof_data {
-            ProofDataPerExpectedInvalidBundle::OutOfRangeTx => BundleValidity::Invalid(
-                InvalidBundleType::OutOfRangeTx(self.mismatched_extrinsic_index),
-            ),
-        };
-
-        expected_bundle_validity == inboxed_bundle_entry.bundle
-    }
-}
-
-/// Fraud proof indicating that a bundle included in `inboxed_bundles` field has incorrect
-/// `bundle_validity` field
-#[derive(Debug, Decode, Encode, TypeInfo, PartialEq, Eq, Clone)]
-pub enum InvalidBundlesFraudProof {
-    TrueInvalid(TrueInvalidBundleEntryFraudProof),
-    FalseInvalid(FalseInvalidBundleEntryFraudProof),
-}
-
-impl InvalidBundlesFraudProof {
-    pub fn domain_id(&self) -> DomainId {
-        match self {
-            InvalidBundlesFraudProof::TrueInvalid(proof) => proof.domain_id,
-            InvalidBundlesFraudProof::FalseInvalid(proof) => proof.domain_id,
-        }
-    }
+    pub fn is_valid_with_context_of_bundle_entry(
```

Comment:
It is hard to go through this function TBH, I believe what this function should do is check the fraud proof against the `inboxed_bundle_entry` to see if it is aligned with `find_inboxed_bundles_mismatch`, this function seems problematic, and doesn't cover all the cases. The following is what I proposed let me know if I miss anything:

```rust
/// Check whether the fraud proof is valid to target the given inboxed bundle from the bad receipt
pub fn is_valid_target_bundle(&self, bad_receipt_bundle: &InboxedBundle) -> bool {
    let invalid_type_of_proof = match self.proof_data {
        ProofDataPerInvalidBundleType::OutOfRangeTx => BundleValidity::Invalid(
            InvalidBundleType::OutOfRangeTx(self.mismatched_extrinsic_index),
        ),
    };
    if !self.is_true_invalid_fraud_proof {
        // `FalseInvalid`
        // The proof trying to prove `bad_receipt_bundle`'s `invalid_bundle_type` is wrong,
        // so the proof should contains the same `invalid_bundle_type`
        bad_receipt_bundle.bundle == BundleValidity::Invalid(invalid_type_of_proof)
    } else {
        // `TrueInvalid`
        match bad_receipt_bundle.bundle {
            // The proof trying to prove the bundle is invalid due to `invalid_type_of_proof` while `bad_receipt_bundle`
            // think it is valid
            BundleValidity::Valid(_) => true,
            BundleValidity::Invalid(invalid_type) => {
                // The proof trying to prove there is an invalid extrinsic that the `bad_receipt_bundle` think is valid,
                // so the proof should point to an extrinsic that in front of the `bad_receipt_bundle`'s
                invalid_type_of_proof.extrinsic_index() < invalid_type.extrinsic_index() ||
                // The proof trying to prove the invalid extrinsic can not pass a check that the `bad_receipt_bundle` think it can,
                // so the proof should point to the same extrinsic and a check that perform before the `bad_receipt_bundle`'s
                (invalid_type_of_proof.extrinsic_index() == invalid_type.extrinsic_index() 
                    && invalid_type_of_proof.checking_order() < invalid_type.checking_order())
            }
        }
    }
}
```

Reply:
@NingLin-P I have refactored this function. And make it more clearer. This is what the body looks like at the moment: 

```rust
    // If this is true, this would mean actual invalid bundle entry in bad receipt is marked as either valid or has different invalid type thn expected.
    // If this is false, this would mean actual valid bundle entry in bad receipt is marked as invalid with specific invalid bundle type.
    let is_true_invalid_proof = invalid_bundle_fraud_proof.is_true_invalid_fraud_proof;

    let is_expected = match (
        true_invalid_bundle_entry.is_invalid(),
        is_true_invalid_proof,
    ) {
        // If entry is not invalid and is true invalid proof then bundle entry is valid while it should be invalid
        (false, true) => true,
        // If entry is invalid and is false invalid proof then bundle entry is invalid while it should be valid
        (true, false) => true,
        // If entry is invalid and is true proof then both are invalid in that case we need to compare further
        (true, true) => {
            // this bundle validity is not expected, because if that is the case then what
            // fraud proof is trying to prove and what the execution receipt says match for this
            // bundle. In that case, the fraud proof is not valid.
            let not_expected_bundle_validity = match invalid_bundle_fraud_proof.proof_data {
                ProofDataPerInvalidBundleType::OutOfRangeTx => {
                    BundleValidity::Invalid(InvalidBundleType::OutOfRangeTx(
                        invalid_bundle_fraud_proof.mismatched_extrinsic_index,
                    ))
                }
            };

            not_expected_bundle_validity != true_invalid_bundle_entry.bundle
        }
        // If entry is not invalid and this is false invalid proof then there is no need for the fraud proof as both conclude to valid bundle entry.
        (false, false) => false,
    };
```

Reply:
It is the refactored version of this func that I commented on, and I don't think it is valid:
- `(false, true) => true,`, this branch is valid
- `(true, false) => true,`, this branch should check `bad_receipt_bundle.bundle == BundleValidity::Invalid(invalid_type_of_proof)`
- `(true, true) => { .. }`, this branch is invalid and it should consider the `extrinsic_index` and `checking_order`
- `(false, false) => false,`, this branch is valid

You can go through my proposal to see what necessary check has been missing. BTW, `match (bool, bool)` really hurt readability.

Reply:
@NingLin-P Okay. While I am still not 100% convinced that either implementation (yours or mine) is correct yours seems to be more intuitive so I have integrated it. 

In the 2nd part when I will be writing generation and test, will correct any issue that arises in this method.